src/cpp/src/sampler.cpp

```diff
@@ -755,7 +755,8 @@ SamplerOutput Sampler::sample(std::vector<SequenceGroup::Ptr> & sequence_groups,
             continue;
 
         size_t num_running_sequences = sequence_group->num_running_seqs();
-        size_t actual_seq_len = sequence_group->get_num_scheduled_tokens(); // points to a token which needs to be sampled
+        size_t actual_seq_len = sequence_group->is_matmul_sliced() ?
```

Comment:
Minor: we can use only `sequence_group->get_seq_len_to_sample()` without any conditions

Reply:
I suppose here we are trying to cover both cases when Gather transformation may not work and has not inserted any Gather.

In this case I think we can just take max(sequence_group->get_num_tokens_to_validate() + 1, sequence_group->get_num_scheduled_tokens())

- first argument covers general sampling of one token and several tokens from speculative decoding. @iefode could you please confirm it's correct? In this case, I hope, we don't need `get_seq_len_to_sample` at all
- second condition imply case w/o gather and a number of scheduled tokens is always at least the same as number of sampled tokens.